core/src/include/ModelArray.hpp

```diff
@@ -0,0 +1,456 @@
+     */
+    ModelArray& operator=(const double& val);
+
+    // ModelArray arithmetic
+    //! Returns a ModelArray containing the per-element sum of the
+    //! object and the provided ModelArray.
+    ModelArray operator+(const ModelArray&) const;
+    //! Returns a ModelArray containing the per-element difference between the
+    //! object and the provided ModelArray.
+    ModelArray operator-(const ModelArray&) const;
+    //! Returns a ModelArray containing the per-element product of the
+    //! object and the provided ModelArray.
+    ModelArray operator*(const ModelArray&) const;
+    //! Returns a ModelArray containing the per-element ratio between the
+    //! object and the provided ModelArray.
+    ModelArray operator/(const ModelArray&) const;
+    // Returns a ModelArray containing the element-wise negation of this.
+    ModelArray operator-() const;
+
+    //! Returns a ModelArray with a constant added to every element of the object.
+    ModelArray operator+(const double&) const;
+    //! Returns a ModelArray with a constant subtracted from every element of the object.
+    ModelArray operator-(const double&) const;
+    //! Returns a ModelArray with every element of the object multiplied by a constant.
+    ModelArray operator*(const double&) const;
+    //! Returns a ModelArray with every element of the object divided by a constant.
+    ModelArray operator/(const double&) const;
+
+    typedef std::vector<size_t> Dimensions;
+
+    //! Returns the number of dimensions of this type of ModelArray.
+    size_t nDimensions() const { return nDimensions(type); }
+    //! Returns the number of dimensions of the specified type of ModelArray.
+    static size_t nDimensions(Type type) { return m_dims.at(type).size(); }
+    //! Returns a vector<size_t> of the size of each dimension of this type of ModelArray.
+    const Dimensions& dimensions() const { return dimensions(type); }
+    //! Returns a vector<size_t> of the size of each dimension of the specified type of ModelArray.
+    static const Dimensions& dimensions(Type type) { return m_dims.at(type); }
+    //! Returns the total number of elements of this type of ModelArray.
+    size_t size() const { return size(type); }
+    //! Returns the total number of elements of the specified type of ModelArray.
+    static size_t size(Type type) { return m_sz.at(type); }
+    //! Returns the size of the data array of this object.
+    size_t trueSize() const { return m_data.rows(); }
+
+    //! Returns a read-only pointer to the underlying data buffer.
+    const double* getData() const { return m_data.data(); }
+    //! Retuns the (enum of) the ModelArray::Type of this.
+    const Type getType() const { return type; }
+
+    /*!
+     * @brief Sets the number and size of the dimensions of a specified type of
+     * ModelArray.
+     *
+     * @param type The type of array the dimensions are to be specified for.
+     * @param dim The per-dimension size to be set.
+     */
```

Comment:
What is the motivation for this being static? Will all model arrays of each type have the same dimensions?